Support sorted writes in the Iceberg connector

alexjo2144 opened the pull request:
## Description

Support sorting files during inserts to the Iceberg connector. This reuses the SortingFileWriter from the Hive connector.

## Non-technical explanation

Sorting enables better performance during selective read queries, where a small range of values is needed from a high carnality column.

## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
# Iceberg
* Support for Iceberg table sort orders. Tables can be created or altered to add a list of `sorted_by` columns which will be used to order files written to the table.
```

alexjo2144 commented:
Note to self: rebase the "Allow updating the sorted_by Iceberg table property" commit to give @osscm author credit.

findinpath commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/SortingFileWriterConfig.java:
```diff
@@ -0,0 +1,60 @@
+        return writerSortBufferSize;
+    }
+
+    @Config("hive.writer-sort-buffer-size")
```
Pre-existing: It would be beneficial to have the purpose of this property documented (in the code).

findinpath commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergV2.java:
```diff
@@ -354,7 +356,7 @@ public void testUpdatingAllTableProperties()
         assertTrue(table.spec().isUnpartitioned());
 
-        assertUpdate("ALTER TABLE " + tableName + " SET PROPERTIES format_version = 2, partitioning = ARRAY['regionkey'], format = 'PARQUET'");
+        assertUpdate("ALTER TABLE " + tableName + " SET PROPERTIES format_version = 2, partitioning = ARRAY['regionkey'], format = 'PARQUET', sorted_by = ARRAY['comment']");
```
Can you showcase a test containing transforms on the sort columns as well?

findinpath commented:
Add compatibility tests with spark.

findinpath commented:
> Sorting enables better performance during selective read queries, where a small range of values is needed from a high cardinality column.

I'd appreciate having a demo test that emphasises on the fact that the number of files being read is smaller when working with `sorted_by`.  Source of inspiration: `io.trino.plugin.iceberg.TestIcebergMetadataFileOperations`

alexjo2144 commented:
Still working on test cases and Marius' comments, but added support for sorting during updates and during `optimize`.

findinpath commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergParquetConnectorTest.java:
```diff
@@ -73,4 +84,26 @@ protected Session withSmallRowGroups(Session session)
+    void verifyFileIsSorted(String path, String sortColumnName)
+    {
+        Comparable previousMax = null;
+        try (ParquetFileReader parquetReader = ParquetFileReader.open(HadoopInputFile.fromPath(new Path(path), ConfigurationInstantiator.newEmptyConfiguration()))) {
```
Naive question: Why not read all the rows from the given parquet file and actually verify that any given row follows the sort order contract in respect to the previous row?

alexjo2144 replied:
Hmmm, we could it just sounds more complex to set up the test.

findinpath commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:
```diff
@@ -1048,6 +1048,176 @@ public void testCreatePartitionedTableWithQuotedIdentifierCasing(String columnNa
+        dropTable("test_blank_sorted_table");
+    }
+
+    abstract void verifyFileIsSorted(String path, String sortColumnName);
```
nit: Can be made `protected`

alexjo2144 commented:
@findinpath @findepi @ebyhr I think this is ready for review. Please take a look when you get a chance

findinpath commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSortingFileWriter.java:
```diff
@@ -0,0 +1,114 @@
+            List<SortOrder> sortOrders,
+            PageSorter pageSorter,
+            TypeOperators typeOperators,
+            SortingFileWriter.TempFileSinkFactory tempFileSinkFactory)
```
`SortingFileWriter.TempFileSinkFactory tempFileSinkFactory` this is an implementation detail.
Initialize directly the `tempFileSinkFactory` while creating the `SortingFileWriter` and remove this constructor parameter.

findinpath commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:
```diff
@@ -1143,6 +1143,69 @@ public void testEmptySortedByList()
+    protected abstract boolean isFileSorted(String path, String sortColumnName);
+
+    @Test
+    public void testSortedNationTable()
```
could you pls add an extra test for sort order evolution:

create table t (a integer, b integer ) with (sort_order=array['a']);
insert into t values (1,100), (3, 3), (2,20)
alter table t set properties sorted_by = array['b'];
insert into t values ....

verify that the first file is sorted by "a" and that the second file is sorted by "b".

findinpath reviewed (commented):
LGTM % comments

findinpath commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:
```diff
@@ -379,7 +384,12 @@ else if (identity.getId() == TRINO_MERGE_PARTITION_DATA) {
-                maxOpenPartitions);
+                maxOpenPartitions,
+                table.getSortOrder(),
+                DataSize.of(100, DataSize.Unit.MEGABYTE),
```
left-over from testing ? 

I assume we would need here `SortingFileWriterConfig`

alexjo2144 replied:
🤦 yeah thanks for catching that. I hardcoded these before I had the config wired up.

alexjo2144 commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergConfig.java:
```diff
@@ -336,4 +337,17 @@ public IcebergConfig setMaterializedViewsStorageSchema(String materializedViewsS
+        return sortedWritingEnabled;
+    }
+
+    @Config("iceberg.sorted-writing-enabled")
```
It can also be useful if your writes are very small (streaming ingest, for example) such that sorting them would be a waste of time until they are compacted.

alexjo2144 commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/SortFields.java:
```diff
@@ -0,0 +1,109 @@
+
+    private static void parseSortField(SortOrderBuilder<?> builder, String field)
+    {
+        boolean matched = tryMatch(field, IDENTITY_ASC_NULLS_FIRST_PATTERN, match -> builder.asc(fromIdentifierToColumn(match.group(1).trim()), NullOrder.NULLS_FIRST)) ||
```
Probably, but I'd rather leave it until we add transform support and clean it up then. I think there are a few other things that should be central between both partition and sort transform parsing.

https://github.com/trinodb/trino/issues/15088

findinpath commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSink.java:
```diff
@@ -286,7 +318,40 @@ private int[] getWriterIndexes(Page page)
+                        sortingFileWriterBufferSize,
+                        sortingFileWriterMaxOpenFiles,
+                        columnTypes,
+                        sortColumnIndexes.build(),
```
Out of scope: `sortColumnIndexes` and `sortOrders` are specified separately even though they very well relate to each other. Ideally we'd have a record which encapsulates them.

findinpath commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSink.java:
```diff
@@ -286,7 +318,40 @@ private int[] getWriterIndexes(Page page)
+            ImmutableList.Builder<Integer> sortColumnIndexes = ImmutableList.builder();
+            ImmutableList.Builder<SortOrder> sortOrders = ImmutableList.builder();
+            for (TrinoSortField sortField : sortOrder) {
+                Types.NestedField icebergType = outputSchema.findField(sortField.getSourceColumnId());
```
`icebergType` -> `column`

findinpath commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergTableHandle.java:
```diff
@@ -44,6 +44,7 @@
     private final Optional<Long> snapshotId;
     private final String tableSchemaJson;
     // Empty means the partitioning spec is not known (can be the case for certain time travel queries).
+    private final List<TrinoSortField> sortOrder;
```
The comment above `sortOrder` corresponds to the field `partitionSpecJson`.

findinpath commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestSortFields.java:
```diff
@@ -0,0 +1,97 @@
+        assertParse("  comment   DESC    NULLS   LAST   ", sortOrder(builder -> builder.desc("comment", NullOrder.NULLS_LAST)));
+        assertParse("comment DESC NULLS LAST", sortOrder(builder -> builder.desc("comment")));
+        assertParse("    comment     DESC   NULLS    LAST   ", sortOrder(builder -> builder.desc("comment")));
+    }
```
Can we add also a few negative tests for transforms to showcase that they are currently not supported in Trino?

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java:
```diff
@@ -200,6 +205,13 @@ public static Map<String, Object> getIcebergTableProperties(Table icebergTable)
         }
 
+        SortOrder sortOrder = icebergTable.sortOrder();
+        // TODO: Support sort column transforms https://github.com/trinodb/trino/issues/15088
```
( https://github.com/trinodb/trino/issues/15088 )

findepi reviewed (commented):
"Support setting the sorted_by property when creating Iceberg tables" LGTM

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/SortFields.java:
```diff
@@ -0,0 +1,109 @@
+        }
+    }
+
+    public static <B extends SortOrderBuilder<?>> void parseSortFields(B sortOrderBuilder, List<String> fields)
```
redundant generics

findepi reviewed (commented):
"Support setting the sorted_by property when creating Iceberg tables"

findepi commented:
@alexjo2144 can you please rebase before (and separate from) applying comments?

findinpath commented on testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergSparkCompatibility.java:
```diff
@@ -1154,6 +1154,38 @@ private void assertSelectsOnSpecialCharacters(String trinoTableName, String spar
+        Assertions.assertThat((String) onTrino().executeQuery("SHOW CREATE TABLE " + trinoTableName).getOnlyValue())
+                .doesNotContain("sorted_by");
+
+        onTrino().executeQuery("INSERT INTO " + trinoTableName + " VALUES (1, 2, 3), (3, 2, 1), (NULL, NULL, NULL)");
```
From https://iceberg.apache.org/spec/#sorting

> A data or delete file is associated with a sort order by the sort order’s id within [a manifest](https://iceberg.apache.org/spec/#manifests). Therefore, the table must declare all the sort orders for lookup.

Could you please inspect the metadata file written by Trino to ensure that it doesn't have any `sort-orders` ? 
I think a combination of:
-  retrieving the metadata file (see   https://github.com/trinodb/trino/blob/390c094273d20f54749fee737b9dbd4377e906e0/testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergSparkCompatibility.java#L2431-L2433 )
- retrieving the metadata file in memory
```
ByteArrayOutputStream bos = new ByteArrayOutputStream();
hdfsClient.loadFile(source + "/" + file, bos);
```
-  construct the table metadata `org.apache.iceberg.TableMetadataParser#fromJson(java.lang.String)` 
 
may help to do this.

alexjo2144 replied:
Looked a little closer at this, I don't think that field is set on the manifest even if we do sort the file. I'm missing a line here:  https://github.com/trinodb/trino/blob/master/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java#L806
```.withSortOrder(sortOrder)```

I wonder what this field is actually used for, though. Maybe Spark avoids sorting data if it's already sorted?

findepi replied:
> I wonder what this field is actually used for, though. Maybe Spark avoids sorting data if it's already sorted?

like https://github.com/trinodb/trino/pull/6634?

>  I don't think that field is set on the manifest even if we do sort the file. 

we definitely need to be storing the sorting information somewhere, otherwise how would the future writes know they should sort?

alexjo2144 replied:
> we definitely need to be storing the sorting information somewhere, otherwise how would the future writes know they should sort?

We do set the `default-sort-order-id` which tells future writes which sort order they should use.

The field Marius is talking about is attached to the individual [Manifests](https://iceberg.apache.org/spec/#manifests). I think it is okay for us not to be setting this value for now.

findinpath commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:
```diff
@@ -1061,17 +1062,17 @@ public void testSortByWithPartitioning()
-
-        dropTable("test_sorted_with_partition_table");
+                assertUpdate("INSERT INTO " + tableName + " VALUES (true, 1, 5), (false, 2, 4), (true, 3, 3), (false, 4, 2), (true, 5, 1)", 5);
+        assertQuery("SELECT * FROM " + tableName, "VALUES (true, 1, 5), (false, 2, 4), (true, 3, 3), (false, 4, 2), (true, 5, 1)");
```
new reminder :)

findepi commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:
```diff
@@ -1219,35 +1256,29 @@ public void testFileSortingWithLargerTable()
+                {"x", "\"x\""},
+                {"X", "\"x\""},
+                {"\"x\"", "\"x\""},
+                {"\"X\"", "\"x\""},
```
The test coverage is missing negative test cases.
What is rejected today, what will be supported in the future?

Also, using `COL` or `NAME` instead of `X` would make it easier to read.
Small x and uppercase X are quite similar, especially when decorated with many quotation marks. and backslashes.

findepi commented:
https://github.com/trinodb/trino/pull/14891#discussion_r1045824548 is just a small nit.

Can you please squash & rebase?

alexjo2144 commented:
Squashed and rebased. Comments all applied except for https://github.com/trinodb/trino/pull/14891#discussion_r1040892611

alexjo2144 commented:
Added docs

findinpath commented on docs/src/main/sphinx/connector/iceberg.rst:
```diff
@@ -712,6 +714,41 @@ In this example, the table is partitioned by the month of ``order_date``, a hash
+    WITH (partitioning = ARRAY['month(order_date)'], sorted_by = ARRAY['order_date'])
+
+If you would like to disable sorting for a particular operation, you can do so by setting the
+session property ``sorted_writing_enabled`` to ``false``;
```
replace `;` with `.`

alexjo2144 replied:
Java muscle memory I guess, thanks

alexjo2144 commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/SortFields.java:
```diff
@@ -0,0 +1,109 @@
+    private static final Pattern IDENTITY_ASC_NULLS_LAST_PATTERN = Pattern.compile(IDENTITY_WHITESPACE_PADDED + "\\s*?(ASC)\\s*?(NULLS\\s*?LAST)\\s*?");
+    private static final Pattern IDENTITY_DESC_PATTERN = Pattern.compile(IDENTITY_WHITESPACE_PADDED + "\\s*?(DESC)\\s*?");
+    private static final Pattern IDENTITY_DESC_NULLS_FIRST_PATTERN = Pattern.compile(IDENTITY_WHITESPACE_PADDED + "\\s*?(DESC)\\s*?(NULLS\\s*?FIRST)\\s*?");
+    private static final Pattern IDENTITY_DESC_NULLS_LAST_PATTERN = Pattern.compile(IDENTITY_WHITESPACE_PADDED + "\\s*?(DESC)\\s*?(NULLS\\s*?LAST)\\s*?");
```
Switched to use antlr

findinpath commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:
```diff
@@ -1063,6 +1064,250 @@ public void testCreatePartitionedTableWithQuotedIdentifierCasing(String columnNa
+        assertUpdate("INSERT INTO " + tableName + " VALUES " + values + ", " + highValues + ", " + lowValues, 3);
+        dropTable(tableName);
+    }
+
```
```
trino:default> CREATE TABLE my_table2 (
            ->         c1 integer,
            ->         c2 date,
            ->         c3 double
            ->     )
            ->     WITH (
            ->         format = 'PARQUET',
            ->         partitioning = ARRAY['c1', 'c2'],
            ->         sorted_by = ARRAY['c4']
            ->     );
Query 20221222_101514_00001_qebin failed: Cannot find field 'c4' in struct: struct<1: c1: optional int, 2: c2: optional date, 3: c3: optional double>
org.apache.iceberg.exceptions.ValidationException: Cannot find field 'c4' in struct: struct<1: c1: optional int, 2: c2: optional date, 3: c3: optional double>
```

```
trino:default> CREATE TABLE my_table2 (
            ->         c1 integer,
            ->         c2 date,
            ->         c3 double
            ->     )
            ->     WITH (
            ->         format = 'PARQUET',
            ->         partitioning = ARRAY['missingpartitionfield'] );
Query 20221222_101645_00003_qebin failed: Unable to parse partitioning value: Cannot find source column: missingpartitionfield
io.trino.spi.TrinoException: Unable to parse partitioning value: Cannot find source column: missingpartitionfield
```

The checks regarding the specified `partitioning` , `sorted_by` field are made only when passing the command to Iceberg framework even though they are actually semantic exceptions that may be caught at Trino level. See the following example where the catching of the exception happens at Trino level:

```
trino:default> CREATE TABLE my_table2 (
            ->         c1 integer,
            ->         c1 date
            ->     )
            ->     WITH (
            ->         format = 'PARQUET');
Query 20221222_102120_00014_qebin failed: line 3:9: Column name 'c1' specified more than once
io.trino.spi.TrinoException: line 3:9: Column name 'c1' specified more than once
	at io.trino.sql.analyzer.SemanticExceptions.semanticException(SemanticExceptions.java:48)
	at io.trino.sql.analyzer.SemanticExceptions.semanticException(SemanticExceptions.java:43)
	at io.trino.execution.CreateTableTask.internalExecute(CreateTableTask.java:170)
```

I see this issue rather as minor/ pre-existing in case of `partitioning` and can be definitely done as a follow-up.

findinpath commented on plugin/trino-iceberg/src/main/antlr4/io/trino/plugin/iceberg/sort/SortDefinition.g4:
```diff
@@ -0,0 +1,60 @@
+ * limitations under the License.
+ */
+
+grammar SortDefinition;
```
👍  for introducing the grammar

findinpath commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/BaseTrinoCatalogTest.java:
```diff
@@ -169,6 +174,71 @@ public void testCreateTable()
         }
     }
 
+    @Test
```
```suggestion
@Test
    public void testCreateWithSortTable()
            throws Exception
    {
        TrinoCatalog catalog = createTrinoCatalog(false);
        String namespace = "test_create_sort_table_" + randomNameSuffix();
        String table = "tableName";
        SchemaTableName schemaTableName = new SchemaTableName(namespace, table);
        catalog.createNamespace(SESSION, namespace, defaultNamespaceProperties(namespace), new TrinoPrincipal(PrincipalType.USER, SESSION.getUser()));
        try {
            Schema tableSchema = new Schema(Types.NestedField.of(1, true, "col1", Types.LongType.get()),
                    Types.NestedField.of(2, true, "col2", Types.StringType.get()),
                    Types.NestedField.of(3, true, "col3", Types.TimestampType.withZone()),
                    Types.NestedField.of(4, true, "col4", Types.StringType.get()));

            SortOrder sortOrder = SortOrder.builderFor(tableSchema)
                    .asc("col1")
                    .desc("col2", NullOrder.NULLS_FIRST)
                    .desc("col3")
                    .desc(Expressions.year("col3"), NullOrder.NULLS_LAST)
                    .desc(Expressions.month("col3"), NullOrder.NULLS_FIRST)
                    .asc(Expressions.day("col3"), NullOrder.NULLS_FIRST)
                    .asc(Expressions.hour("col3"), NullOrder.NULLS_FIRST)
                    .desc(Expressions.bucket("col2", 10), NullOrder.NULLS_FIRST)
                    .desc(Expressions.truncate("col4", 5), NullOrder.NULLS_FIRST).build();
            catalog.newCreateTableTransaction(
                            SESSION,
                            schemaTableName,
                            tableSchema,
                            PartitionSpec.unpartitioned(),
                            sortOrder,
                            arbitraryTableLocation(catalog, SESSION, schemaTableName),
                            ImmutableMap.of())
                    .commitTransaction();
            assertThat(catalog.listTables(SESSION, Optional.of(namespace))).contains(schemaTableName);
            assertThat(catalog.listTables(SESSION, Optional.empty())).contains(schemaTableName);

            Table icebergTable = catalog.loadTable(SESSION, schemaTableName);
            assertEquals(icebergTable.name(), quotedTableName(schemaTableName));
            assertEquals(icebergTable.schema().columns().size(), 4);
            assertEquals(icebergTable.schema().columns().get(0).name(), "col1");
            assertEquals(icebergTable.schema().columns().get(0).type(), Types.LongType.get());
            assertEquals(icebergTable.schema().columns().get(1).name(), "col2");
            assertEquals(icebergTable.schema().columns().get(1).type(), Types.StringType.get());
            assertEquals(icebergTable.schema().columns().get(2).name(), "col3");
            assertEquals(icebergTable.schema().columns().get(2).type(), Types.TimestampType.withZone());
            assertEquals(icebergTable.schema().columns().get(3).name(), "col4");
            assertEquals(icebergTable.schema().columns().get(3).type(), Types.StringType.get());
            assertEquals(icebergTable.sortOrder(), sortOrder);

            catalog.dropTable(SESSION, schemaTableName);
        }
        finally {
            if (!catalog.listTables(SESSION, Optional.of(schemaTableName.getSchemaName())).isEmpty()) {
                catalog.dropTable(SESSION, schemaTableName);
            }
            catalog.dropNamespace(SESSION, namespace);
        }
    }
```


Test is failing otherwise when working with Hive metastore

```
2023-01-06T17:25:56.4534678Z [ERROR] io.trino.plugin.iceberg.catalog.hms.TestTrinoHiveCatalogWithHiveMetastore.testCreateWithSortTable  Time elapsed: 3.776 s  <<< FAILURE!
2023-01-06T17:25:56.4535707Z java.lang.IllegalArgumentException: java.net.UnknownHostException: hadoop-master
2023-01-06T17:25:56.4537083Z 	at org.apache.hadoop.security.SecurityUtil.buildTokenService(SecurityUtil.java:447)
2023-01-06T17:25:56.4537804Z 	at org.apache.hadoop.hdfs.NameNodeProxiesClient.createProxyWithClientProtocol(NameNodeProxiesClient.java:131)
2023-01-06T17:25:56.4538407Z 	at org.apache.hadoop.hdfs.DFSClient.<init>(DFSClient.java:355)
2023-01-06T17:25:56.4538816Z 	at org.apache.hadoop.hdfs.DFSClient.<init>(DFSClient.java:289)
2023-01-06T17:25:56.4539331Z 	at org.apache.hadoop.hdfs.DistributedFileSystem.initialize(DistributedFileSystem.java:176)
2023-01-06T17:25:56.4539910Z 	at io.trino.hdfs.TrinoFileSystemCache.createFileSystem(TrinoFileSystemCache.java:167)
2023-01-06T17:25:56.4540454Z 	at io.trino.hdfs.TrinoFileSystemCache.getInternal(TrinoFileSystemCache.java:123)
2023-01-06T17:25:56.4540960Z 	at io.trino.hdfs.TrinoFileSystemCache.get(TrinoFileSystemCache.java:91)
2023-01-06T17:25:56.4541473Z 	at org.apache.hadoop.fs.ForwardingFileSystemCache.get(ForwardingFileSystemCache.java:38)
2023-01-06T17:25:56.4541961Z 	at org.apache.hadoop.fs.FileSystem.get(FileSystem.java:479)
2023-01-06T17:25:56.4542369Z 	at org.apache.hadoop.fs.Path.getFileSystem(Path.java:361)
2023-01-06T17:25:56.4542815Z 	at io.trino.hdfs.HdfsEnvironment.lambda$getFileSystem$0(HdfsEnvironment.java:74)
2023-01-06T17:25:56.4543348Z 	at io.trino.hdfs.authentication.NoHdfsAuthentication.doAs(NoHdfsAuthentication.java:25)
2023-01-06T17:25:56.4543871Z 	at io.trino.hdfs.HdfsEnvironment.getFileSystem(HdfsEnvironment.java:73)
2023-01-06T17:25:56.4544330Z 	at io.trino.hdfs.HdfsEnvironment.getFileSystem(HdfsEnvironment.java:67)
2023-01-06T17:25:56.4544851Z 	at io.trino.filesystem.hdfs.HdfsFileSystem.deleteDirectory(HdfsFileSystem.java:116)
2023-01-06T17:25:56.4545735Z 	at io.trino.plugin.iceberg.catalog.AbstractTrinoCatalog.deleteTableDirectory(AbstractTrinoCatalog.java:185)
2023-01-06T17:25:56.4546415Z 	at io.trino.plugin.iceberg.catalog.hms.TrinoHiveCatalog.dropTable(TrinoHiveCatalog.java:321)
2023-01-06T17:25:56.4547108Z 	at io.trino.plugin.iceberg.catalog.BaseTrinoCatalogTest.testCreateWithSortTable(BaseTrinoCatalogTest.java:232)
2023-01-06T17:25:56.4547757Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2023-01-06T17:25:56.4548338Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
2023-01-06T17:25:56.4548977Z 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2023-01-06T17:25:56.4549509Z 	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
2023-01-06T17:25:56.4550013Z 	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
2023-01-06T17:25:56.4550522Z 	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
2023-01-06T17:25:56.4550959Z 	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
2023-01-06T17:25:56.4551417Z 	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
2023-01-06T17:25:56.4551917Z 	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
2023-01-06T17:25:56.4552431Z 	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
2023-01-06T17:25:56.4552951Z 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
2023-01-06T17:25:56.4553577Z 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
2023-01-06T17:25:56.4554006Z 	at java.base/java.lang.Thread.run(Thread.java:833)
2023-01-06T17:25:56.4554566Z Caused by: java.net.UnknownHostException: hadoop-master
2023-01-06T17:25:56.4554848Z 	... 32 more
```

findinpath commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/sort/CaseInsensitiveStream.java:
```diff
@@ -0,0 +1,91 @@
+        int result = stream.LA(i);
+
+        switch (result) {
+            case 0:
```
nit: we can make use of an enhanced switch here.

findinpath commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:
```diff
@@ -787,13 +795,36 @@ private IcebergWritableTableHandle newWritableTableHandle(SchemaTableName name,
                 retryMode);
     }
 
+    private List<TrinoSortField> getSortFieldsIfSupported(SortOrder sortOrder)
```
nit: method can be static

homar commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSink.java:
```diff
@@ -286,7 +318,40 @@ private int[] getWriterIndexes(Page page)
+                    .map(partition -> locationProvider.newDataLocation(partitionSpec, partition, fileName))
+                    .orElseGet(() -> locationProvider.newDataLocation(fileName));
+
+            ImmutableList.Builder<Integer> sortColumnIndexes = ImmutableList.builder();
```
Would it make sense to perform all these sorted related operations only if sorting is actually enabled?

homar commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSinkProvider.java:
```diff
@@ -122,7 +143,12 @@ public ConnectorPageSink createPageSink(ConnectorTransactionHandle transactionHa
                         optimizeHandle.getTableStorageProperties(),
-                        maxOpenPartitions);
+                        maxOpenPartitions,
+                        optimizeHandle.getSortOrder(),
```
I see you pass these 3 fields in many places, maybe it would make sense to create an object (a record) to represent them ? Similar to SortingFileWriterConfig ? WDYT ?

homar commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/SortFields.java:
```diff
@@ -0,0 +1,162 @@
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public final class SortFields
```
Would it be possible to add a small javadoc here explaining purpose of this class?

alexjo2144 replied:
Gonna rename to `SortFieldUtils`, I think that makes more sense.

homar commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/SortFields.java:
```diff
@@ -0,0 +1,162 @@
+        }
+    }
+
+    public static void parseSortFields(SortOrderBuilder<?> sortOrderBuilder, List<String> fields)
```
TBH I think that this method is not needed. It doesn't do much and is only used in a method above(if this is not correct then please ignore this comment). Also because it is overloaded it makes the picture less clear at least for me

alexjo2144 replied:
It's used one other place, in the `IcebergMetadata` code that changes the sort order when you use `SET PROPERTIES`

homar commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:
```diff
@@ -1063,6 +1064,250 @@ public void testCreatePartitionedTableWithQuotedIdentifierCasing(String columnNa
     }
 
+    @Test
+    public void testSortByAllTypes()
```
Could you please add a test with UPDATE, DELETE and MERGE? They should work but just in case

findepi commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/SortingFileWriterConfig.java:
```diff
@@ -0,0 +1,63 @@
+    }
+
+    @LegacyConfig("hive.max-open-sort-files")
+    @Config("sorting.writer.max-open-files")
```
I am not convinced about the `sorting.` prefix.
Sorting feels too granular for a top level name.

- you can leave `hive.` prefix for these configs, we can defer the decision this way
- or, you can remove `hive.` from this class, and bind it with "hive" prefix in Hive and "iceberg" prefix in Iceberg. (config binder allows prefixes for contextual binding)

findepi reviewed (commented):
"Extract writer sorting properties from HiveConfig"

findepi commented on docs/src/main/sphinx/connector/iceberg.rst:
```diff
@@ -748,6 +750,41 @@ In this example, the table is partitioned by the month of ``order_date``, a hash
     WITH (partitioning = ARRAY['month(order_date)', 'bucket(account_number, 10)', 'country'])
 
+Sorted tables
+------------------
```
@mosabua do we expect the header section underline to be same length in characters as the header section title itself?

mosabua replied:
yes

findepi commented on docs/src/main/sphinx/connector/iceberg.rst:
```diff
@@ -748,6 +750,41 @@ In this example, the table is partitioned by the month of ``order_date``, a hash
+
+In addition to partitioning, Iceberg tables may have a specified list of sort columns. When writes
+are made to the table, each file will individually be sorted by these columns. Sorting can be
+particularly helpful when selective reads are frequently made against a column with a high cardinality.
```
"against" or sth like "with predicates on [columns table is sorted on]" ?
(my wording isn't good, that's just for conveying the intention)

findepi commented on docs/src/main/sphinx/connector/iceberg.rst:
```diff
@@ -748,6 +750,41 @@ In this example, the table is partitioned by the month of ``order_date``, a hash
+are made to the table, each file will individually be sorted by these columns. Sorting can be
+particularly helpful when selective reads are frequently made against a column with a high cardinality.
+
+Sort orders are provided using the ``sorted_by`` table property. For example::
```
"Sorting order is" ?

findepi commented on docs/src/main/sphinx/connector/iceberg.rst:
```diff
@@ -748,6 +750,41 @@ In this example, the table is partitioned by the month of ``order_date``, a hash
+        account_number BIGINT,
+        customer VARCHAR,
+        country VARCHAR)
+    WITH (partitioning = ARRAY['country'], sorted_by = ARRAY['order_date'])
```
If partitioning isn't strictly required for sorting, let's have the first example not use it.

mosabua replied:
agreed

findepi commented on docs/src/main/sphinx/connector/iceberg.rst:
```diff
@@ -748,6 +750,41 @@ In this example, the table is partitioned by the month of ``order_date``, a hash
+        country VARCHAR)
+    WITH (partitioning = ARRAY['country'], sorted_by = ARRAY['order_date'])
+
+Sorting can be combined with partitioning on the same column, for example if a timestamp
+column is often queried selectively but the data is sparse enough that a higher granularity
+partition transform produces a better data layout::
```
Is it worth clarifying that we don't mean "identity partitioning" (so not "any partitioning") here?

"data is sparse enough" -- i don't understand the condition
if data is very sparse, you won't have many rows per partition, and sorting won't help
would "is dense enough" be considered appropriate?

mosabua replied:
in either case "enough" is not really a concrete value .. more like an estimate .. we should try to be more definitive.

alexjo2144 replied:
> if data is very sparse, you won't have many rows per partition, and sorting won't help

That's the condition I was trying to describe. A timestamp column is sparse enough that you want to lump en entire year worth of data into a partition but predicates are still using day or hour precision.

mosabua replied:
Do we even need to explain when this is beneficial .. maybe not .. maybe we can just say that you can do it. And bring the example ..

findepi commented on docs/src/main/sphinx/connector/iceberg.rst:
```diff
@@ -748,6 +750,41 @@ In this example, the table is partitioned by the month of ``order_date``, a hash
+session property ``sorted_writing_enabled`` to ``false``.
+
+While the Iceberg specification supports the same set of transforms on sort columns as it does on
+partition columns, only sorting by the column's raw value is currently supported in the connector.
```
Is it worth adding what Trino does when writing to a table with such unsupported sorting?
Does it fail, or ignore sorting?

(it should ignore)

mosabua replied:
I feel like these two sentences are not needed or should be simplified. 

Maybe

Sorting is only performed based on the raw values of the configured columns.

alexjo2144 replied:
> Sorting is only performed based on the raw values of the configured columns.

That's not quite correct, at least as I read it. If you have a sort configuration of `ARRAY[truncate(a, 2), b]` we will sort the data by `b`, not `a, b`. Does that make sense?

alexjo2144 replied:
I have updated those sentences though, hope that's better.

findepi reviewed (commented):
"Document Iceberg sorted_by table property"

findepi commented on plugin/trino-iceberg/src/main/antlr4/io/trino/plugin/iceberg/sort/SortDefinition.g4:
```diff
@@ -0,0 +1,54 @@
+}
+
+sortDefinition
+    : identifier direction=(ASCENDING | DESCENDING)? (NULLS nullOrder=(FIRST | LAST))? EOF;
```
This can be easily parsed with a simple regex.
I've added a fixup commit proposing such a change, please take a look

findepi reviewed (commented):
"Support the Iceberg sorted_by table property"

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergTableHandle.java:
```diff
@@ -360,6 +375,7 @@ public boolean equals(Object o)
                 tableType == that.tableType &&
                 Objects.equals(snapshotId, that.snapshotId) &&
                 Objects.equals(tableSchemaJson, that.tableSchemaJson) &&
+                Objects.equals(sortOrder, that.sortOrder) &&
```
add to hashCode too

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergWritableTableHandle.java:
```diff
@@ -57,6 +59,7 @@ public IcebergWritableTableHandle(
         this.schemaAsJson = requireNonNull(schemaAsJson, "schemaAsJson is null");
         this.partitionsSpecsAsJson = ImmutableMap.copyOf(requireNonNull(partitionsSpecsAsJson, "partitionsSpecsAsJson is null"));
         this.partitionSpecId = partitionSpecId;
+        this.sortOrder = ImmutableList.copyOf(requireNonNull(sortOrder, "sortOrderId is null"));
```
update the message to match the var name

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TrinoSortField.java:
```diff
@@ -0,0 +1,83 @@
+    public int hashCode()
+    {
+        return Objects.hash(sourceColumnId, sortOrder);
+    }
```
add toString too

findepi commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:
```diff
@@ -1174,7 +1175,6 @@ public void testEmptySortedByList()
         dropTable(tableName);
     }
 
-
```
belongs to ``Support the Iceberg `sorted_by` table property`` commit

(https://github.com/trinodb/trino/pull/15730)

findepi commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:
```diff
@@ -1220,6 +1220,94 @@ public static Object[][] sortedTableWithSortTransform()
         };
     }
 
+    protected abstract boolean isFileSorted(String path, String sortColumnName);
```
move test helper below tests using it

findepi commented on testing/trino-faulttolerant-tests/src/test/java/io/trino/faulttolerant/iceberg/TestIcebergParquetFaultTolerantExecutionConnectorTest.java:
```diff
@@ -68,6 +68,13 @@ public void testStatsBasedRepartitionDataOnInsert()
+    protected boolean isFileSorted(String path, String sortColumnName)
+    {
+        // This test produces many small files compared to the superclass. These files do not have enough data to test ordering using row group statistics.
+        throw new SkipException("Unimplemented");
```
That's too late to skip the test: we already spent time writing data. 
(I'd rather throw Unsupported here, and override each of the tests to disable it -- if disabling is what i want)

Also, small file still can be sorted or not, so writing small files should prevent us from verifying sortedness.

BTW do you know why this class writes many small files? in its setup i don't see anything about it (did i miss?), so it feels like FTE implicitly results in many small files, implying FTE is not ready for a show yet (many small files is a big problem!)
cc @losipiuk @arhimondr

alexjo2144 replied:
> Also, small file still can be sorted or not, so writing small files should prevent us from verifying sortedness.

We could. In the other classes I'm just checking row group min/max values but we could read each row

alexjo2144 replied:
> BTW do you know why this class writes many small files?

I do not

findepi commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergAvroConnectorTest.java:
```diff
@@ -49,4 +49,10 @@ public void testIncorrectIcebergFileSizes()
+    @Override
+    protected boolean isFileSorted(String path, String sortColumnName)
+    {
+        throw new SkipException("Unimplemented");
```
see comment in FTE test. throw unsupported and override selected methods

also, worth having a TODO + issue?

findepi replied:
or, maybe sorting Avro files has no sense (has it?), so we should skip doing that.
why add latency for the writer, if the reader not gonna leverage that?

findepi commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:
```diff
@@ -1220,6 +1220,94 @@ public static Object[][] sortedTableWithSortTransform()
+        String tableName = "test_sorted_nation_table_" + randomNameSuffix();
+        assertUpdate(withSmallRowGroups, "CREATE TABLE " + tableName + " WITH (sorted_by = ARRAY['comment']) AS SELECT * FROM nation", 25);
+        for (Object filePath : computeActual("SELECT file_path from \"" + tableName + "$files\"").getOnlyColumnAsSet()) {
+            assertTrue(isFileSorted((String) filePath, "comment"));
```
do we have a test sorting desc
do we have a test sorting by two columns
and covering values containing nulls?

alexjo2144 replied:
All are covered by the SparkCompatibilityTests, but I can duplicate them here too if you'd like.

findepi commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergOrcConnectorTest.java:
```diff
@@ -56,6 +69,42 @@ protected Session withSmallRowGroups(Session session)
+            String previousMax = null;
+            OrcColumnId sortColumnId = orcReader.getRootColumn().getNestedColumns().stream()
+                    .filter(column -> column.getColumnName().equals(sortColumnName))
+                    .findFirst()
+                    .orElseThrow()
```
You don't want first, you want the only.

```
.collect(onlyElement())
```

findepi commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergOrcConnectorTest.java:
```diff
@@ -56,6 +69,42 @@ protected Session withSmallRowGroups(Session session)
+                    .orElseThrow()
+                    .getColumnId();
+            List<StripeStatistics> statistics = orcReader.getMetadata().getStripeStatsList().stream()
+                    .filter(Optional::isPresent)
```
why?

alexjo2144 replied:
Rather than `get()` or `orElseThrow()`?

findepi commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergOrcConnectorTest.java:
```diff
@@ -56,6 +69,42 @@ protected Session withSmallRowGroups(Session session)
+                // TODO: This only works if the sort column is a String
+                StringStatistics columnStatistics = stripeStatistics.getColumnStatistics().get(sortColumnId).getStringStatistics();
+
+                if (previousMax != null && previousMax.compareTo(columnStatistics.getMin().toStringUtf8()) > 0) {
```
My IDE displays a `Method invocation 'toStringUtf8' may produce 'NullPointerException' ` warning here.
I know it doesn't matter, but I also don't like leaving warnings behind, since it makes me less likely notice important warnings next time.

findepi commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergOrcConnectorTest.java:
```diff
@@ -56,6 +69,42 @@ protected Session withSmallRowGroups(Session session)
+            throw new UncheckedIOException(e);
+        }
+
+        return true;
```
move inside the `try`, right after the `for (StripeStatistics stripeStatistics : statistics)` loop

findepi commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergParquetConnectorTest.java:
```diff
@@ -73,4 +83,29 @@ protected Session withSmallRowGroups(Session session)
+            for (BlockMetaData blockMetaData : parquetReader.getRowGroups()) {
+                ColumnChunkMetaData columnMetadata = blockMetaData.getColumns().stream()
+                        .filter(column -> getOnlyElement(column.getPath().iterator()).equalsIgnoreCase(sortColumnName))
+                        .findFirst()
```
`.collect(onlyElement())`

findepi commented on testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergSparkCompatibility.java:
```diff
@@ -1162,6 +1162,38 @@ private void assertSelectsOnSpecialCharacters(String trinoTableName, String spar
+        Assertions.assertThat((String) onTrino().executeQuery("SHOW CREATE TABLE " + trinoTableName).getOnlyValue())
+                .contains("sorted_by = ARRAY['b ASC NULLS FIRST','c DESC NULLS LAST']");
+
+        onTrino().executeQuery("INSERT INTO " + trinoTableName + " VALUES (1, 2, 3), (3, 2, 1), (NULL, NULL, NULL)");
+    }
```
INSERT and no verification?
maybe skip the insert if you don't care about it results.

(or add a comment why it's a good idea to test that insert does not fail, but i think we're well covered by Iceberg BCT)

alexjo2144 replied:
I meant to come back and finish these tests but totally forgot...

findepi commented on testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergSparkCompatibility.java:
```diff
@@ -1162,6 +1162,38 @@ private void assertSelectsOnSpecialCharacters(String trinoTableName, String spar
+        onSpark().executeQuery("ALTER TABLE " + sparkTableName + " WRITE ORDERED BY truncate(b, 3)");
+
+        Assertions.assertThat((String) onTrino().executeQuery("SHOW CREATE TABLE " + trinoTableName).getOnlyValue())
+                .doesNotContain("sorted_by");
```
Let's add a test case showing that the supported prefix (suffix too?) sort fields are retained.

Is it worth adding a test case showing that changing some other table attribute (eg column type) doesn't erase sort ordering that Trino couldn't understand?

findepi commented:
Overall LGTM, see previous comments.

Also see @homar 's comments.

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSink.java:
```diff
@@ -286,7 +318,40 @@ private int[] getWriterIndexes(Page page)
+                        columnTypes,
+                        sortColumnIndexes.build(),
+                        sortOrders.build(),
+                        pageSorter,
```
Trino sorting may or may not be same as Iceberg sorting, at least it's not per Iceberg spec.
Actually. I don't think there is any problem with _the code_ here, but I wish Iceberg spec got this fixed: https://github.com/apache/iceberg/pull/2891
cc @RussellSpitzer @yyanyy

mosabua commented on docs/src/main/sphinx/connector/iceberg.rst:
```diff
@@ -748,6 +750,41 @@ In this example, the table is partitioned by the month of ``order_date``, a hash
+Sorted tables
+------------------
+
+In addition to partitioning, Iceberg tables may have a specified list of sort columns. When writes
```
Yeah ... remove that section .. the two are unrelated

alexjo2144 commented:
Just rebasing so far, had conflicts with the change for update/delete to use merge machinery, but it's simpler now.

findinpath commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSink.java:
```diff
@@ -286,7 +318,40 @@ private int[] getWriterIndexes(Page page)
+            ImmutableList.Builder<SortOrder> sortOrders = ImmutableList.builder();
+            for (TrinoSortField sortField : sortOrder) {
+                Types.NestedField column = outputSchema.findField(sortField.getSourceColumnId());
+                sortColumnIndexes.add(outputSchema.columns().indexOf(column));
```
Could you pls verify whether `outputSchema.columns()` lookup still works in the context of dealing with nested types?

alexjo2144 replied:
Good catch, did not think about sorting on non-base columns.

alexjo2144 commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSink.java:
```diff
@@ -135,6 +157,16 @@ public IcebergPageSink(
+        this.sortedWritingEnabled = isSortedWritingEnabled(session);
+        this.sortingFileWriterBufferSize = requireNonNull(sortingFileWriterBufferSize, "sortingFileWriterBufferSize is null");
+        this.sortingFileWriterMaxOpenFiles = sortingFileWriterMaxOpenFiles;
+        this.tempDirectory = new Path(locationProvider.newDataLocation("trino-tmp-files"));
```
> Is it ok that every query will use the same directory under data/?

I think so? Seemed like we would want them to be somewhere where they'll get cleaned up by GC.

> who's creating that directory?

Good catch, I only tested this on S3 but it'll probably fail on Azure.

> we're not deleting it, right? (necessary, if shared)

We are not

alexjo2144 replied:
> Good catch, I only tested this on S3 but it'll probably fail on Azure.

I take that back, testing on Azure today and ABFS did not mind if I create the directory or not.

alexjo2144 replied:
I'll add a test for that after we merge: https://github.com/trinodb/trino/pull/16032

alexjo2144 commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSink.java:
```diff
@@ -286,7 +318,40 @@ private int[] getWriterIndexes(Page page)
+            ImmutableList.Builder<SortOrder> sortOrders = ImmutableList.builder();
+            for (TrinoSortField sortField : sortOrder) {
+                Types.NestedField column = outputSchema.findField(sortField.getSourceColumnId());
+                sortColumnIndexes.add(outputSchema.columns().indexOf(column));
```
I'll at least move it to only be computed once per PageSink rather than once per writer, but `n` and `m` should be small.

alexjo2144 commented:
Comments applied, thanks for the reviews all.

alexjo2144 commented on testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergSparkCompatibility.java:
```diff
@@ -1238,13 +1238,18 @@ public void testTrinoReadsSparkSortOrder()
 
-        onTrino().executeQuery("INSERT INTO " + trinoTableName + " VALUES (1, 2, 3), (3, 2, 1), (NULL, NULL, NULL)");
+        onTrino().executeQuery("INSERT INTO " + trinoTableName + " VALUES (3, 2, 1), (1, 2, 3), (NULL, NULL, NULL)");
+        assertThat(onSpark().executeQuery("SELECT _pos, a, b, c FROM " + sparkTableName))
```
It is the position of the row in the file: https://github.com/apache/iceberg/blob/master/core/src/main/java/org/apache/iceberg/MetadataColumns.java#L38

alexjo2144 commented on testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergSparkCompatibility.java:
```diff
@@ -1255,12 +1260,17 @@ public void testTrinoIgnoresUnsupportedSparkSortOrder()
                 .doesNotContain("sorted_by");
 
         onTrino().executeQuery("INSERT INTO " + trinoTableName + " VALUES (1, 2, 3), (3, 2, 1), (NULL, NULL, NULL)");
+        assertThat(onSpark().executeQuery("SELECT _pos, a, b, c FROM " + sparkTableName))
+                .contains(
+                        row(0, 1, 2, 3),
+                        row(1, 3, 2, 1),
```
There are two sort fields, the first is a `truncate` that we ignore, but the second one we apply

findepi commented:
LGTM, but not to the CI.

Please squash commits to satisfy the check-commit check and force push. Add a comment.
if any fixes needed (per maven checks), force push again

alexjo2144 commented:
`testFileSortingWithLargerTable` is causing problems with heap size, should have expected that. I'm going to reduce `writer-sort-buffer-size` for the tests and try to force a file system flush of the sorting writers with a smaller table.

alexjo2144 commented:
Alright, I think I fixed that.

@mosabua I think we're ready for you to take a look at the documentation when you get a chance.

mosabua commented on docs/src/main/sphinx/connector/iceberg.rst:
```diff
@@ -776,6 +778,43 @@ In this example, the table is partitioned by the month of ``order_date``, a hash
+
+A table's sort order is provided using the ``sorted_by`` property. For example::
+
+    CREATE TABLE iceberg.testdb.customer_orders (
```
example.customers.orders

mosabua commented on docs/src/main/sphinx/connector/iceberg.rst:
```diff
@@ -776,6 +778,43 @@ In this example, the table is partitioned by the month of ``order_date``, a hash
+column is often queried selectively but the data is sparse enough that a higher granularity
+partition transform produces a better data layout::
+
+    CREATE TABLE iceberg.testdb.customer_orders (
```
example.customers.orders

mosabua commented on docs/src/main/sphinx/connector/iceberg.rst:
```diff
@@ -776,6 +778,43 @@ In this example, the table is partitioned by the month of ``order_date``, a hash
+        account_number BIGINT,
+        customer VARCHAR,
+        country VARCHAR)
+    WITH (partitioning = ARRAY['month(order_date)'], sorted_by = ARRAY['order_date'])
```
format over two lines

mosabua commented on docs/src/main/sphinx/connector/iceberg.rst:
```diff
@@ -776,6 +778,43 @@ In this example, the table is partitioned by the month of ``order_date``, a hash
+        country VARCHAR)
+    WITH (partitioning = ARRAY['month(order_date)'], sorted_by = ARRAY['order_date'])
+
+If you would like to disable sorting for a particular operation, you can do so by setting the
```
You can disable sorted writing for with the session property ``sorted_writing_enabled`` set to ``false``.

However .. is that a catalog session property?

mosabua reviewed (changes requested):
Added a whole bunch of suggestions for the docs. Let me know when its ready to review again @alexjo2144

alexjo2144 commented:
Thanks @mosabua comment applied

mosabua commented on docs/src/main/sphinx/connector/iceberg.rst:
```diff
@@ -776,6 +778,46 @@ In this example, the table is partitioned by the month of ``order_date``, a hash
+Sorting is particularly beneficial when the sorted columns show a high cardinality
+and are used as a filter for selective reads.
+
+The sort order is configured with the sorted by table property.
```
```suggestion
The sort order is configured with the ``sorted`` by table property.
```

mosabua commented on docs/src/main/sphinx/connector/iceberg.rst:
```diff
@@ -776,6 +778,46 @@ In this example, the table is partitioned by the month of ``order_date``, a hash
+The sort order is configured with the sorted by table property.
+Specify an array of one or more columns to use for sorting
+when creating the table. The following example configures the
+order_date column in the orders table in the customers schema in the example catalog::
```
```suggestion
``order_date`` column of the ``orders`` table in the ``customers`` schema in the ``example`` catalog::
```

mosabua commented on docs/src/main/sphinx/connector/iceberg.rst:
```diff
@@ -776,6 +778,46 @@ In this example, the table is partitioned by the month of ``order_date``, a hash
+        sorted_by = ARRAY['order_date']
+    )
+
+You can disable sorted writing with the session property sorted_writing_enabled set to false.
```
```suggestion
You can disable sorted writing with the catalog session property ``sorted_writing_enabled`` set to ``false``.
```

mosabua commented on docs/src/main/sphinx/connector/iceberg.rst:
```diff
@@ -776,6 +778,46 @@ In this example, the table is partitioned by the month of ``order_date``, a hash
+
+You can disable sorted writing with the session property sorted_writing_enabled set to false.
+
+While the Iceberg specification supports a number of transforms on sort columns, they are
```
I would remove that .. we try to document only what works .. not all the different exceptions and limitations..

mosabua replied:
Also .. what do you mean .. how could you even have a transforms .. use a transforming function in the table property? Would that not just fail with a syntax error?

alexjo2144 replied:
You can define them in Spark

mosabua replied:
I would still remove that..

mosabua reviewed (changes requested):
Second doc review done .

alexjo2144 commented:
Second pass of doc review, applied

losipiuk commented on testing/trino-faulttolerant-tests/src/test/java/io/trino/faulttolerant/iceberg/TestIcebergParquetFaultTolerantExecutionConnectorTest.java:
```diff
@@ -68,11 +69,52 @@ public void testStatsBasedRepartitionDataOnInsert()
+    @Test
+    public void testSortedNationTable()
+    {
+        throw new SkipException("File sorting for parquet files is handled by TestIcebergParquetConnectorTest");
```
> FTE shouldn't result in much smaller files than non-FTE run, since tables with large number of (small) files are expensive to query. In an ideal world there shouldn't be such limitation

This should not be the case. Are you observing it? Actually I run the test with FTE and I got just one file and it passed just fine.

> it would be good to test Iceberg write-time sorting with FTE. For example, if we implemented sorting using SortNode in the plan, non-FTE would get data sorted and FTE wouldn't (am i right here @losipiuk @arhimondr )

If you would depend on the global sorting mechanism in Trino - then yeah - it would not work out of the box with FTE because of [this](https://github.com/trinodb/trino/blob/920d14c82c5adefd2f06feadfac1a0a96f9ff5ac/core/trino-main/src/main/java/io/trino/SystemSessionProperties.java#L1198-L1201)

alexjo2144 commented:
@mosabua did you want to take another pass on the docs?

mosabua commented on docs/src/main/sphinx/connector/iceberg.rst:
```diff
@@ -783,6 +785,46 @@ In this example, the table is partitioned by the month of ``order_date``, a hash
+Sorted tables
+-------------
+
+The connector supports sorted files as a performance improvement. Data is sorted during writes
```
Can you fix wrapping in there paragraphs to be 80 char width?

alexjo2144 commented:
Addressed doc comments.

I do need to fix the tests though. I moved some to the smoke tests so that they run against all supported file systems but they're not working yet.

alexjo2144 commented:
@ebyhr @findepi I think I got everything green with this one. Can you kick off a build with serets?

ebyhr commented:
/test-with-secrets sha=0f09920b81b690612b026fcfd6e7c4cb252951ee

findepi commented:
Run with secrets failed: https://github.com/trinodb/trino/issues/13199 (reopened)

findepi commented:
rebased to resolve a conflict

findepi commented:
thanks!
